vm/ubpf_vm.c

```diff
@@ -293,6 +293,17 @@ i32(uint64_t x)
     return x;
 }
 
+/**
+ * @brief Sign extend immediate value to a signed 64-bit value.
+ *
+ * @param[in] immediate The signed 32-bit immediate value to sign extend.
+ * @return The sign extended 64-bit value.
+ */
+static int64_t sign_extend_immediate(int32_t immediate) {
+    return (int64_t)immediate;
```

Comment:
I am currently working through whether or not this is guaranteed to give the proper behavior according to the C standard. In newer versions of the C standard, twos complement representation of signed integers is specified. In older versions, it is not. Therefore, we will (I think) need to be very careful about this. See, e.g., https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2218.htm.

Reply:
Further research:

1. GCC represents integers as twos complement: https://gcc.gnu.org/onlinedocs/gcc-5.3.0/gcc/Integers-implementation.html#Integers-implementation
2. MSVC represents integers as twos complement: https://learn.microsoft.com/en-us/cpp/c-language/range-of-integer-values?view=msvc-170

There does not appear to be a place where Clang (llvm) defines its implementation-specific behavior. See https://github.com/llvm/llvm-project/issues/11644 for the status of such a document.

Reply:
In the yet-to-be standardized version of C, twos complement is specified as the representation to use for signed integers. See https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf